### PR TITLE
adding an action to the router contract

### PR DESCRIPTION
**Pull request by sogipec**

### CODE REVIEW

File: contracts/BaseRouter.sol [hunk 1]
```diff
@@ -78,7 +78,8 @@ enum ActionType {
     openPerpetual,
     addToPerpetual,
     veANGLEDeposit,
-    claimRewardsWithPerps
+    claimRewardsWithPerps,
+    addCollateral
```
Comment (GuillaumeNervoXS) [inline contracts/BaseRouter.sol:82]: make it explicit that it is linked to borrow module `addCollateralBorrower` or so

File: contracts/BaseRouter.sol [hunk 2]
```diff
@@ -387,6 +394,25 @@ abstract contract BaseRouter is Initializable {
...
+    function _addCollateral(
+        address vaultManager,
+        uint256 vaultID,
+        uint256 amount
+    ) internal virtual {
+        ActionBorrowType[] memory action = new ActionBorrowType[](1);
+        action[0] = ActionBorrowType.addCollateral;
+        bytes[] memory dataBorrow = new bytes[](1);
+        dataBorrow[0] = abi.encode(vaultID, amount);
+        IVaultManagerFunctions(vaultManager).angle(action, dataBorrow, address(0), address(0), address(0), "0x");
```
Comment (GuillaumeNervoXS) [inline contracts/BaseRouter.sol:413]: if you want to have an empty byte you should just do `bytes memory emptyByte;`

File: contracts/BaseRouter.sol [hunk 3]
```diff
@@ -487,15 +513,12 @@ abstract contract BaseRouter is Initializable {
     /// @param to Address to which shares should be sent
     /// @param maxAmountIn Max amount of assets used to mint
     /// @return amountIn Amount of assets used to mint by `to`
-    function _mintSavingsRate(
+    function _mint4626(
         IERC4626 savingsRate,
         uint256 shares,
         address to,
         uint256 maxAmountIn
     ) internal returns (uint256 amountIn) {
```
Comment (GuillaumeNervoXS) [inline contracts/BaseRouter.sol:521]: put my comment in @dev please

File: contracts/BaseRouter.sol [hunk 4]
```diff
@@ -232,7 +233,14 @@ abstract contract BaseRouter is Initializable {
                 _parseVaultIDs(actionsBorrow, dataBorrow, vaultManager);
                 _changeAllowance(IERC20(collateral), address(vaultManager), type(uint256).max);
                 _angleBorrower(vaultManager, actionsBorrow, dataBorrow, to, who, repayData);
-                _changeAllowance(IERC20(collateral), address(vaultManager), 0);
+            } else if (actions[i] == ActionType.addCollateralBorrower) {
+                (address collateral, address vaultManager, uint256 vaultID, uint256 amount) = abi.decode(
+                    data[i],
+                    (address, address, uint256, uint256)
+                );
+                if (amount == type(uint256).max) amount = IERC20(collateral).balanceOf(address(this));
```
Comment (greedythib) [inline contracts/BaseRouter.sol:241]: assuming all sweep actions went fine and there are not collat on the router, the caller has to do a transfer before calling this action

**Review by GuillaumeNervoXS — APPROVED** [review]

**Pull request merged**